### DISCUSSION TITLE
Using two hubs in a robotics event

### DISCUSSION THREAD

**Discussion by Attilio-Galzerano** [pybricks/support#1202] [Q&A] [answered]
I'm a coach for a robotics team that will use pybricks to compete in a local event. We are planning to use in our robot two spike prime hubs to communicate each other via bluetooth. I gotta say that me and my students are loving pybricks, thank you all for developing this amazing project! So, I have some doubts and some concerns that I would love to get some light on and to see if the way I'm thinking is correct.

- If the code is already downloaded in the hub, changing the battery will not remove the code. During the event the students can safely change batteries and won't need to download the code again to that hub?

- The web IDE will work just fine, even without internet? they will be able to flash the lego firmware or the pybricks firmware if needed, change the code, connect via bluetooth? No need to have internet, correct?

- In regards to the color sensor of the spike prime collection, every time that I use the function color(), reflection() or hsv() it already uses the full brightness of all of its leds, right? No need to turn the sensor's lights on before measuring something. Also, the default behavior is: lights on with full brightness all the time?

- One thing we will try to do in our competition is identify a certain silver tape with our color sensor, in order to do this, I'm thinking it is going to be necessary to calibrate the hsv values for this color and to add it to the detectable colors of each sensor separately. This code bellow is how we are currently trying to solve this part, is this the correct way? Also to be more precise we will need to do this for each color sensor, if we use more than one for this purpose, right?

```python
from pybricks.hubs import PrimeHub
from pybricks.pupdevices import Motor, ColorSensor, UltrasonicSensor, ForceSensor
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch


hub = PrimeHub(broadcast_channel=50, observe_channels=[55])

leftLineSensor = ColorSensor(Port.E)
centerColorSensor = ColorSensor(Port.C)
rightLineSensor = ColorSensor(Port.A)


leftMotor = Motor(Port.D, positive_direction=Direction.COUNTERCLOCKWISE)
rightMotor = Motor(Port.F)


drive_base = DriveBase(leftMotor, rightMotor, wheel_diameter=32, axle_track=198)


distanceSensor = UltrasonicSensor(Port.B)


def main():
    leftHue = leftSaturation = leftValue = 0
    rightHue = rightSaturation = rightValue = 0

    pressed = []

    
    while Button.LEFT not in pressed:
        pressed = hub.buttons.pressed()
        wait(10)
    
    
    while any (hub.buttons.pressed()):
        wait(10)
    
    hub.speaker.beep()
    wait(1000)

    for i in range(20):
        leftColor = leftLineSensor.hsv()
        rightColor = rightLineSensor.hsv()

        a = leftColor.h
        b = leftColor.s
        c = leftColor.v

        x = rightColor.h
        y = rightColor.s
        z = rightColor.v

        

        leftHue += a
        leftSaturation += b
        leftValue += c

        rightHue += x
        rightSaturation += y
        rightValue += z

    i += 1 # to calculate the average from the samples
    leftHue /= i
    leftSaturation /= i
    leftValue /= i

    rightHue /= i
    rightSaturation /= i
    rightValue /= i

    Color.myGrayLeft = Color(h=leftHue, s=leftSaturation, v=leftValue)
    Color.myGrayRight = Color(h=rightHue, s=rightSaturation, v=rightValue)

    myLeftColors = [Color.BLACK, Color.WHITE, Color.GREEN, Color.RED, Color.myGrayLeft]
    myRightColors = [Color.BLACK, Color.WHITE, Color.GREEN, Color.RED, Color.myGrayRight]

    leftLineSensor.detectable_colors(myLeftColors)
    rightLineSensor.detectable_colors(myRightColors)

    print(leftLineSensor.detectable_colors())
    print(rightLineSensor.detectable_colors())

    while True:

        print("=====================================")
        print("Cores:", leftLineSensor.color(), " | ", rightLineSensor.color())
        print("HSV: ", leftLineSensor.hsv(), " | ", rightLineSensor.hsv())
        print()
        wait(1000)
        


try:
    main()
finally:
    centerColorSensor.lights.off()
    rightLineSensor.lights.off()
    leftLineSensor.lights.off()`
```
- The other thing that I'm worried about is the bluetooth communication between hubs at the competition. To establish a bluetooth connection to a lego spike prime hub (that has pybricks firmware installed) is only possible via the pybricks web IDE? some other team at the event using a lego official IDE or other programm will not be able to see or connect accidentally to our hubs? Let's say that in the event we are the only team using pybricks, does this helps in a way that no one will be able to connect accidentally to our hubs? during a robot round, we will be using the bluetooth broadcast and observe between our two hubs, can anything interfere or cause trouble to this communication? Is there something we can do to protect against this? 

- We are currently using the broadcast and observe in two ways, one of them is to constantly broadcast and observe in a forever loop, the other way is inside a function to broadcast only if some circumstance happens - are there any problems with this way? is there a better way of doing this? Calling the broadcast once makes the hub broadcast constantly the message until i call the broadcast again, or it broadcasts the message only once and I risk the other hub miss the message? Below are the two hub codes that we are trying to implement the bluetooth communications

the main hub has this code:
```python
from pybricks.hubs import PrimeHub
from pybricks.pupdevices import Motor, ColorSensor, UltrasonicSensor, ForceSensor
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch


hub = PrimeHub(broadcast_channel=50, observe_channels=[55])


leftLineSensor = ColorSensor(Port.E)
centerColorSensor = ColorSensor(Port.C)
rightLineSensor = ColorSensor(Port.A)


leftMotor = Motor(Port.D, positive_direction=Direction.COUNTERCLOCKWISE)
rightMotor = Motor(Port.F)


drive_base = DriveBase(leftMotor, rightMotor, wheel_diameter=32, axle_track=198)


distanceSensor = UltrasonicSensor(Port.B)


#a function to find the object and send this info to the other hub
def findObject():

    flagObject = centerColorSensor.color()

    # if it found the object communicate to the other hub
    if flagObject == Color.MAGENTA:
        hub.light.on(Color.MAGENTA)
        drive_base.stop()
        hub.ble.broadcast(1) # 1 = found the object 
        drive_base.straight(-50, wait=True)
        drive_base.turn(170)
        drive_base.straight(-100)
        wait(2000)
        hub.ble.broadcast(0) # send 0 so the other hub doesn't try to grab the object again
        drive_base.turn(170, then=Stop.COAST)
        drive_base.straight(30, then=Stop.COAST)
        
def avoidObstacle():

    distance = distanceSensor.distance()

    if distance < 42:
        drive_base.straight(-50, Stop.COAST)
        drive_base.turn(-90, Stop.COAST)

        drive_base.drive(100, 0)
        #here I need to get the distance values from the other hub sensors
        #to circumvent the obstacle

        while True:
            data = hub.ble.observe(55)

            if data is not None:
                rightDistance = data[1]

                #passed the obstacle
                if rightDistance > 100:
                    drive_base.straight(150, Stop.COAST)
                    hub.speaker.beep()
                    break
        
        drive_base.turn(90, Stop.COAST)
        
        drive_base.drive(100, 0)

        while True:
            data = hub.ble.observe(55)

            if data is not None:
                rightDistance = data[1]

                if rightDistance < 100:
                    hub.speaker.beep()
                    break
        
        while True:
            data = hub.ble.observe(55)

            if data is not None:
                rightDistance = data[1]

                if rightDistance > 100:
                    hub.speaker.beep()
                    break
        
        drive_base.straight(180, Stop.COAST)
        drive_base.turn(90, Stop.COAST)

        drive_base.drive(100, 0)

        while True:
            data = hub.ble.observe(55)

            if data is not None:
                rightDistance = data[1]

                if rightDistance < 100:
                    hub.speaker.beep()
                    break
        
        while leftLineSensor.color() != Color.NONE and rightLineSensor.color() != Color.NONE:
            wait(10)
        
        drive_base.straight(50, Stop.COAST)
        drive_base.turn(-80, Stop.COAST)
        drive_base.turn(-90, Stop.COAST, False)

        while rightLineSensor.color() != Color.NONE:
            wait(10)
        
        drive_base.stop()
        hub.speaker.beep()


def main():

    #go straight forever until find object
    drive_base.drive(100, 0)

    while True:
        findObject()

        avoidObstacle()

        wait(100)

main()
```

the second hub code:
```python
from pybricks.hubs import PrimeHub
from pybricks.pupdevices import Motor, ColorSensor, UltrasonicSensor, ForceSensor
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch


hub = PrimeHub(broadcast_channel=55, observe_channels=[50])

liftMotor = Motor(Port.D, gears=[12, 20])
clawMotor = Motor(Port.F, gears=[1, 24], reset_angle=True)
rightDistanceSensor = UltrasonicSensor(Port.B)
leftDistanceSensor = UltrasonicSensor(Port.C)

def openClaw():
    clawMotor.run_angle(30, 40)
       

def closeClaw():
    clawMotor.run_until_stalled(-30, Stop.HOLD, 40)
    

def main():
      
    closeClaw()
    
        
    while True:
        
        leftDistance = leftDistanceSensor.distance()
        rightDistance = rightDistanceSensor.distance()

        hub.ble.broadcast(leftDistance, rightDistance)

        data = hub.ble.observe(50)

        if data is not None:
            if data[0] == 1:
                # the other hub found the object, grab it
                openClaw()
                wait(3000)
                closeClaw()
                wait(4000)
                    
        wait(10)
    

try:
    main()
finally:
    pass

```

Thanks!

**Comment by laurensvalk** [comment, marked as answer]
> If the code is already downloaded in the hub, changing the battery will not remove the code. During the event the students can safely change batteries and won't need to download the code again to that hub?

Correct!

Caveat: a new program is saved *when the hub shuts down normally*.

So, turn the hub off before you pull the battery out if you just loaded a new program onto it.

> The web IDE will work just fine, even without internet? they will be able to flash the lego firmware or the pybricks firmware if needed, change the code, connect via bluetooth? No need to have internet, correct?

Correct.

Caveat: Everything is stored in your browser. If you empty all your cache/cookies, etc., then you need internet access again to use Pybricks.

This normally doesn't happen automatically, but some school/work computers may erase this.

> In regards to the color sensor of the spike prime collection, every time that I use the function color(), reflection() or hsv() it already uses the full brightness of all of its leds, right? No need to turn the sensor's lights on before measuring something. Also, the default behavior is: lights on with full brightness all the time?

Correct.

> One thing we will try to do in our competition is identify a certain silver tape with our color sensor, in order to do this, I'm thinking it is going to be necessary to calibrate the hsv values for this color and to add it to the detectable colors of each sensor separately. This code bellow is how we are currently trying to solve this part, is this the correct way?

Almost, but not quite. If you're going to change one color, it's better to do all of the colors you need. So Instead of

```python
myLeftColors = [Color.BLACK, Color.WHITE, Color.GREEN, Color.RED, Color.myGrayLeft]
```

It would be better to do

```python
myLeftColors = [Color.myBlack, Color.MyWhite, Color.MyGreen, Color.MyRed, Color.myGrayLeft]
```

> Also to be more precise we will need to do this for each color sensor, if we use more than one for this purpose, right?

Possibly. Perhaps compare how different they are. But perhaps more importantly, make sure to select the values based on your surroundings too. Maybe the competition mat has a different print, or bright sunshine coming through roof windows, etc.

> The other thing that I'm worried about is the bluetooth communication between hubs at the competition. To establish a bluetooth connection to a lego spike prime hub (that has pybricks firmware installed) is only possible via the pybricks web IDE? some other team at the event using a lego official IDE or other programm will not be able to see or connect accidentally to our hubs? Let's say that in the event we are the only team using pybricks, does this helps in a way that no one will be able to connect accidentally to our hubs? 

The LEGO app won't find a hub running Pybricks, but another team that uses the Pybricks website could still find your hub. You could pick a good name for your hub so it's clearly yours.

And it can only be found while the hub light is blinking, so not while your robot is running already.

> during a robot round, we will be using the bluetooth broadcast and observe between our two hubs, can anything interfere or cause trouble to this communication? Is there something we can do to protect against this?

In an event with many robots and mobile phones in the room, there is always the possibility of interference, especially if the robots are far apart with all the other devices in between.

> We are currently using the broadcast and observe in two ways, one of them is to constantly broadcast and observe in a forever loop, the other way is inside a function to broadcast only if some circumstance happens - are there any problems with this way? is there a better way of doing this? Calling the broadcast once makes the hub broadcast constantly the message until i call the broadcast again, or it broadcasts the message only once and I risk the other hub miss the message? Below are the two hub codes that we are trying to implement the bluetooth communications

Once you call `broadcast()`, it will keep broadcasting that message until you broadcast something else, so another hub should also pick it up if you start the program later.

Performance on SPIKE/MINDSTORMS is relatively good. On Technic/City/Move hub, Bluetooth is a bit less reliable, so keep that in mind if you plan to mix hubs.

> I gotta say that me and my students are loving pybricks, thank you all for developing this amazing project! 

Thank you! We'd love to hear how the competition goes or see a video of the robots in action :smile:

**Comment by Attilio-Galzerano** [comment]
The competition happened last Saturday. This is the robot 'Skorpion' that my students made:

[https://imgur.com/a/s2HjNO5](https://imgur.com/a/s2HjNO5)


One problem that we had during the event was that we had a color sensor in front of the robot, pointing ahead of it to detect a blue rescue kit that the robot needed to grab. However, at one point, when the robot was going up a ramp, it kept detecting the blue rescue kit ahead of it when there was nothing there. Since the default setting of the color detection's surface argument is True, I thought this problem of detecting a blue object when there is "nothing" in front of the sensor was not possible. My guess is that the light above the competition table was bluish, and the sensor detected this in the ambient light. Maybe we could have calibrated the color of the blue object on the sensor to avoid this...



Anyway, I want to thank you guys again for the excellent work with Pybricks. We hope to use it again for other competitions.

**Reply by laurensvalk** [reply]
Thanks for sharing, and well done to the students! :smile: 

And yes, color sensors can have a bit of a mind of their own depending on external conditions :)
